teuthology/openstack/__init__.py

```diff
@@ -961,33 +1026,66 @@ def get_user_data(self):
             '--setup-pulpito',          #do_setup_pulpito=true
             '--populate-paddles',       #do_populate_paddles=true
         ]
-        ceph_workbench = ''
+
         if self.args.ceph_workbench_git_url:
             all_options += [
                 '--setup-ceph-workbench',
                 '--ceph-workbench-branch %s' % self.args.ceph_workbench_branch,
                 '--ceph-workbench-git-url %s' % self.args.ceph_workbench_git_url,
             ]
+        if self.args.no_canonical_tags:
+            all_options += [ '--no-canonical-tags' ]
+        if self.args.upload:
+            all_options += [ '--archive-upload ' + self.args.archive_upload ]
+        if network:
+            all_options += [ '--network ' + network ]
+        if self.args.simultaneous_jobs:
+            all_options += [ '--nworkers ' + str(self.args.simultaneous_jobs) ]
+
+        cmds = [
+            "su - -c '(set -x ; %s && cd teuthology && ./bootstrap install)' "
+            "%s >> /tmp/init.out 2>&1" % (clone, self.username),
+            #"echo 'export %s' | tee /home/%s/openrc.sh" % (openrc, self.username),
+            cmd_str(
+                "su - -c 'cat | tee $HOME/openrc.sh' {user} <<EOF\n"
+                "export {openrc}\n"
+                "EOF\n".format(user=self.username, openrc=openrc)),
+            cmd_str(
+                "su - -c '(set -x ; source openrc.sh ; cd teuthology ; "
+                "source virtualenv/bin/activate ; "
+                "teuthology/openstack/setup-openstack.sh {opts})' "
+                "{user} >> /tmp/init.out "
+                "2>&1".format(user=self.username,
+                              opts=' '.join(setup_options + all_options))),
+            "/etc/init.d/teuthology restart"
+        ]
+        if cacert_cmd:
+            cmds.insert(0,cmd_str(cacert_cmd))
+        #cloud-config
+        cloud_config = {
+            'bootcmd': [
+                'touch /tmp/init.out',
+                'echo nameserver 8.8.8.8 | tee -a /etc/resolv.conf',
+            ],
+            'manage_etc_hosts': True,
+            'system_info': {
+                'default_user': {
+                    'name': self.username
+                }
+            },
+            'packages': [
+                'python-virtualenv',
+                'git',
+                'rsync',
+            ],
+            'runcmd': cmds,
+            'final_message': 'teuthology is up and running after $UPTIME seconds'
+        }
+        user_data = "#cloud-config\n%s" % \
+              yaml.dump(cloud_config, default_flow_style = False)
+        open(path, 'w').write(user_data)
```

Comment:
ah, so you are generating the user-data at runtime instead of maintaining it as a separate file in the repo - nice!

Reply:
If we ever try to migrate the teuthology VM to openSUSE or SLE, the user-data will probably need to be different from this, since cloud-init on SUSE does not work the same as on Ubuntu. But that's just a comment, not an argument against generating user-data at runtime.